iothub/device/src/Transport/Mqtt/MqttTransportHandler.cs

```diff
@@ -4,1310 +4,1043 @@
+                switch (connectCode)
+                {
+                    case MqttClientConnectResultCode.BadUserNameOrPassword:
+                    case MqttClientConnectResultCode.NotAuthorized:
+                    case MqttClientConnectResultCode.ClientIdentifierNotValid:
+                        throw new IotHubClientException("Failed to open the MQTT connection due to incorrect or unauthorized credentials", cfe, false, IotHubStatusCode.Unauthorized);
+                    case MqttClientConnectResultCode.UnsupportedProtocolVersion:
+                        // Should never happen since the protocol version (3.1.1) is hardcoded
+                        throw new IotHubClientException("Failed to open the MQTT connection due to an unsupported MQTT version", cfe);
+                    case MqttClientConnectResultCode.ServerUnavailable:
+                        throw new IotHubClientException("MQTT connection rejected because the server was unavailable", cfe, true, IotHubStatusCode.ServerBusy);
+                    default:
+                        // MQTT 3.1.1 only supports the above connect return codes, so this default case
+                        // should never happen. For more details, see the MQTT 3.1.1 specification section "3.2.2.3 Connect Return code"
+                        // https://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html
+                        // MQTT 5 supports a larger set of connect codes. See the MQTT 5.0 specification section "3.2.2.2 Connect Reason Code"
+                        // https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901074
+                        throw new IotHubClientException("Failed to open the MQTT connection", cfe);
+                }
             }
         }
 
-        private async Task WaitUntilC2dMessageArrivesAsync(CancellationToken cancellationToken)
-        {
-            cancellationToken.ThrowIfCancellationRequested();
-            CancellationToken disconnectToken = _disconnectAwaitersCancellationSource.Token;
-            EnsureValidState();
-
-            using var linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, disconnectToken);
-
-            // Wait until either of the linked cancellation tokens have been canceled.
-            await _receivingSemaphore.WaitAsync(linkedCts.Token).ConfigureAwait(false);
-        }
-
-        public override async Task CompleteMessageAsync(string lockToken, CancellationToken cancellationToken)
+        public override async Task SendEventAsync(Message message, CancellationToken cancellationToken)
         {
-            if (Logging.IsEnabled)
-                Logging.Enter(this, $"Completing a message with lockToken: {lockToken}", nameof(CompleteMessageAsync));
-
             cancellationToken.ThrowIfCancellationRequested();
-            EnsureValidState();
 
-            if (_qosReceivePacketFromService == QualityOfService.AtMostOnce)
+            string TopicName;
```

Comment:
string topicName = PopulateMessagePropertiesFromMessage(
    _moduleId == null
        ? _deviceToCloudMessagesTopic
        : _moduleToCloudMessagesTopic,
    message);

Reply:
That's more concise, but I don't love sticking ternary operators within parameter lists

Reply:
I can move the ternary operator above the method call though

Reply:
You could make the call twice in the ternary, or just use that for the parameter and make the call once.